report/test/renderer/components-test.js

```diff
@@ -0,0 +1,80 @@
+/**
+ * @license Copyright 2021 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/* eslint-env jest */
+
+import fs from 'fs';
+import jsdom from 'jsdom';
+import expect from 'expect';
+import {DOM} from '../../renderer/dom.js';
+import {LH_ROOT} from '../../../root.js';
+import {getTextNodePossiblySignificantText} from '../../../build/build-report-components.js';
+
+const html = fs.readFileSync(LH_ROOT + '/report/assets/templates.html', 'utf-8');
+const {window} = new jsdom.JSDOM(html);
+const tmplEls = window.document.querySelectorAll('template');
+
+/**
+ * @param {HTMLTemplateElement} tmplEl
+ */
+async function assertDOMTreeMatches(tmplEl) {
+  global.document = window.document;
+  global.Node = window.Node;
+  global.DocumentFragment = window.DocumentFragment;
+
+  const dom = new DOM(window.document);
+
+  function cleanUselessNodes(parent) {
```

Comment:
it would be great to get some docs on these two functions for future changes folks might have to make. Basically, why things need to get removed or reordered so it's clear what might need to be added.